### ISSUE TITLE
Adhere to logging best practices... after figuring them out

### ISSUE THREAD

**Issue by nathanielmanistaatgoogle** [grpc/grpc#16572]
Our use of `logging` **must**

* not emit garbage "No handlers could be found" messages to any output, ever
* be suitably configurable by applications

. It probably *should*

* be easy enough for us to maintain without asking "what's with [these `basicConfig` calls](https://github.com/grpc/grpc/pull/16378)?"
* adhere to Python community-wide best practices ([what even are those?](https://bugs.python.org/issue34590))

. Some other open questions are

* should we use one logger object, several logger objects in a tree relationship, or several unrelated logger objects?
  * I'm leaning "just one, with the name 'grpc'", but it would be nice to see that inclination endorsed by some kind of guidance from someone in a position to know.
  * We currently use one-per-source file, which is roughly one-per-module. Turns out this isn't necessarily the best way to use logger objects. Who knew?
* should we expose the logger object we use as a library-scope constant or should we just refer to it by name in our documentation?

[This documentation addressing library maintainers](https://docs.python.org/3.7/howto/logging.html#configuring-logging-for-a-library) is authoritative, but it just isn't sufficiently detailed to be helpful. I've filed [this documentation enhancement request](https://bugs.python.org/issue34590) to have it improved.

**Comment by levi-rs** [comment]
The `1.16.0` release, which includes the new `basicConfig` calls, clobbers my logging configurations in the microservices I build.

**Comment by kuzmiigo** [comment]
We have the same problem: 1.6.0 release broke logging in our Python app (no output at all now when INFO level logging was visible before). It seems that logging configuration (e.g. using `logging.basicConfig()`) should be done on the top app level and avoided on a library level.

For example, from the ‘Configuring Logging for a Library’ section of the [Python Logging HOWTO](https://docs.python.org/3/howto/logging.html#configuring-logging-for-a-library):

> It is strongly advised that you _do not add any handlers other than_ `NullHandler` _to your library’s loggers_. This is because the configuration of handlers is the prerogative of the application developer who uses your library. The application developer knows their target audience and what handlers are most appropriate for their application: if you add handlers ‘under the hood’, you might well interfere with their ability to carry out unit tests and deliver logs which suit their requirements.

**Comment by nathanielmanistaatgoogle** [comment]
@levi-rs, @kuzmiigo: with what versions of the Python interpreter are you executing your code and seeing these problems?

**Comment by roshammar** [comment]
I have the same issue, running grpcio version 1.16.0 on Python 3.6.6 on Ubuntu.

Minimal example:

```python
import logging
import grpc

def main():

    logger = logging.getLogger('myapp')
    hdlr = logging.FileHandler('myapp.log')
    logger.addHandler(hdlr)
    logger.setLevel(logging.INFO)

    logger.info('Hello')

if __name__ == '__main__':
    main()
```

With grpc 1.15.0 my log message ends up in the log file only, but using 1.16.0 it also gets printed to `stderr`.

**Comment by kuzmiigo** [comment]
I used Python 3.6.6 on macOS 10.14.

**Comment by levi-rs** [comment]
```
$ docker run --entrypoint python <my docker image>:local -V
Python 3.6.6
```

**Comment by kuzmiigo** [comment]
Here is a simple example step by step (Python 3.6.6 on macOS 10.14).

**Version 1.** Let’s enable logging for INFO level:
```python
import logging

log_fmt = '%(asctime)s %(levelname)s %(message)s'
logging.basicConfig(level=logging.INFO, format=log_fmt)

logger = logging.getLogger(__name__)
logger.info('TEST info')
logger.error('TEST error')
```
Result (as expected):
```
2018-10-26 21:21:11,255 INFO TEST info
2018-10-26 21:21:11,255 ERROR TEST error
```
**Version 2.** Now, let’s import `grpc` (1.16.0):
```python
import grpc
import logging

log_fmt = '%(asctime)s %(levelname)s %(message)s'
logging.basicConfig(level=logging.INFO, format=log_fmt)

logger = logging.getLogger(__name__)
logger.info('TEST info')
logger.error('TEST error')
```
Result:
```
ERROR:__main__:TEST error
```
As you can see, the format has changed and INFO messages do not appear.

**Version 3.** Let’s import `grpc` after we have called `logging.basicConfig()`:
```python
import logging

log_fmt = '%(asctime)s %(levelname)s %(message)s'
logging.basicConfig(level=logging.INFO, format=log_fmt)

import grpc

logger = logging.getLogger(__name__)
logger.info('TEST info')
logger.error('TEST error')
```
Result:
```
2018-10-26 21:27:32,829 INFO TEST info
2018-10-26 21:27:32,829 ERROR TEST error
```
This result is the same as version 1 and is expected. So, it is possible to fix this problem by reordering imports, but it is a hack and it breaks PEP8 compliance.

The reason for this problem is that `logging.basicConfig()` can be called once only. More precisely, it has effect on the first invocation only. Some quotes from docs:

* This function does nothing if the root logger already has handlers configured. (`logging.basicConfig()` docstring)
* As it’s intended as a one-off simple configuration facility, only the first call will actually do anything: subsequent calls are effectively no-ops. ([Logging HOWTO](https://docs.python.org/3/howto/logging.html#logging-to-a-file))

It seems like it is better to call `logging.basicConfig()` once in the `main()`. Libraries should not do that as they will essentially hijack logging configuration from applications that might use them.

**Comment by gnossen** [comment]
This issue was partially addressed https://github.com/grpc/grpc/pull/17064. It seems to put logging back into a reasonable state, but doesn't address Nathan's fundamental question of how we should break logging up in the system.

**Comment by gnossen** [comment]
I'm trying to get a feel for what the rest of the python ecosystem does with its logging, so I looked into the top few libraries on pypi:

### `urllib3`

`urllib3` maintains a logger for not quite every module, but for each one that does heavy lifting. The logger name is `__name__`, and no handlers are registered for any module-level loggers, including `NullHandler`. Their documentation [spells out how to configure logging for the library](https://github.com/requests/requests/blob/ff0c325014f817095de35013d385e137b111d6e8/docs/api.rst#api-changes).

They explicitly configure a library root-level logger called `urllib3` to which they attach a `NullHandler`. This addresses the "no handlers could be found problem".

Their tests explicitly configure handlers, just like ours do.

### `scrapy`

`scrapy` is more hands-on. It provides a [configuration module for its logging](https://github.com/scrapy/scrapy/blob/master/scrapy/utils/log.py) and [a whole document](https://github.com/scrapy/scrapy/blob/master/docs/topics/logging.rst) on how to handle logging with `scrapy`. It looks like `log.py`'s whole reason for existence is making sure that a handler is attached to to the `scrapy` handler at startup.

I think the extra complexity here is because `scrapy` also offers a CLI, so there has to be someway to configure logging without resorting to writing python.

---
 handler with 
Based on all of the libraries I've looked at, I think our current approach is reasonable. The one change I would make is to explicitly configure a `grpc` logger and to only attach a `NullHandler` to it instead of putting the burden on the author of each new module to configure it there.

tl;dr: I think we should
 - Configure a logger in each module that cares about logging
 - Do *not* attach any handler to the module level loggers (including `NullHandler`s)
 - Explicity configure a `grpc` logger with a `NullHandler` attached

**Comment by alertedsnake** [comment]
I've been playing with the 1.16.0 release for a bit, and the logging system annoyingly gets in my way.  I'm using logging.setLoggerClass() to set my default logger class to be one that outputs messages in a specific format, and this runs at the start of the application before all imports, but then every time one of the grpcio packages is imported, it adds a second handler onto the root logger, resulting in duplicate messages for each line.  I have to go and clean that out after all the imports (`logging.root.handlers = []` does the trick) - but it would be nice if the grpc libraries didn't do this in the first place.

**Comment by gnossen** [comment]
@alertedsnake We have fixed this problem. It'll be available soon in a patch to 1.16. (https://github.com/grpc/grpc/issues/17017)

**Comment by alertedsnake** [comment]
@gnossen awesome, thanks!

**Comment by gnossen** [comment]
Resolved by https://github.com/grpc/grpc/pull/17143.
